### ISSUE TITLE
onStart not working

### ISSUE THREAD

**Issue by WilliamHua** [sorich87/bootstrap-tour#395]
What are the conditions for onStart to be called? I am putting a simple alert in the onStart method and it doesn't call when I force start the tour

**Comment by ramatevish** [comment]
onStart should be called called when `tour.start()` is called (with or without force), after step 0 is shown, if the tour has never started before. As you can see below, if `_current` has been set, onStart won't get called.

```
    Tour.prototype.start = function(force) {
      var promise;
      if (force == null) {
        force = false;
      }
      if (!this._inited) {
        this.init(force);
      }
      if (this._current === null) {
        promise = this._makePromise(this._options.onStart != null ? this._options.onStart(this) : void 0);
        this._callOnPromiseDone(promise, this.showStep, 0);
      }
      return this;
    };
```

**Comment by ramatevish** [comment]
It does look like a bug to me after a quick look.

**Comment by WilliamHua** [comment]
Is there anyway to call some JavaScript functions before step 0 is shown, even if the tour has been shown before?

I need to call some JavaScript before the tour even start. I'm having issues where the setup code (which freezes parts of the screen so there isn't any interactivity) is called, but the tour isn't starting. It's okay if the tour doesn't start but I want to make it so that the setup code isn't called unless the tour starts.

Also would you guys have any idea why the tour starts on certain machines but not others?

**Comment by ramatevish** [comment]
First of all, if you *need* this to work (without force), you can change that `===` to `==` and I believe it should work as described. You could call `tour.restart()` to have it re-run. It should re-call onStart.

I generally wrap my tours in another object if I need to handle some setup task, or check for things.

**Comment by WilliamHua** [comment]
My original code was in a wrapper:
        initTour = function(){
                setUp();

                // Instance the tour
                tour = new Tour({
                    // more settings
                }); 
                tour.init();
                tour.setCurrentStep(0);
                tour.start(true);
          }

However, there would be instances where the setUp method would be called yet the tour wouldn't run. I couldn't figure out why the tour didn't run, I could not consistently reproduce it, but there are enough people reporting it for us to believe it's a big enough issue. 

I tried this:

        initTour = function(){
                // Instance the tour
                tour = new Tour({
                    onStart: function(tour) {
                        setUp();
                    },
                    // more settings
                }); 
                tour.init();
                tour.setCurrentStep(0);
                tour.restart();
          }

And then I call the initTour function, however this doesn't work either. This is after I changed the === to == in the standalone code. The problem is that setUp HAS to be called before the tour is started because it is adjusting dynamic elements for the tour to attach to.

**Comment by ramatevish** [comment]
First off, I fixed onStart failure last night, and the proper fix was a bit different then the one I recommended here, so I'd recommend grabbing the latest `bootstrap-tour.js`.

In your code example, instead of calling `setUp()` in `onStart()`, why not call it before tour, then create your array of steps? E.g.

```
initTour = function(){
           var stepsArray = setUp();
            // Instance the tour
            tour = new Tour({
               steps: stepsArray,
                // more settings
            }); 
            tour.init();
            tour.setCurrentStep(0);
            tour.restart();
      }
```

**Comment by ramatevish** [comment]
I fixed this properly last night – see #397. You can get the lastest version of `bootstrap-tour.js` to get that fix. That said, your example code wouldn't be influenced: calling `restart()` would call `onStart` even before the fix (see #396). This sounds like a different issue at this point.

**Comment by WilliamHua** [comment]
Yes you're right. You mentioned that onStart is called once the tour has started, which made me realize that onStart can't be used since my methods have to be called BEFORE the tour has started.

Is there anyway to have a function similar to onStart that is called before the tour has started? Once again, I can't put code into production because there is a bug with the tour not starting 100% of the time. I either need the tour to start 100% of the time (preferable), or I need the setup code to only run when the tour starts.
